### PR TITLE
[WFLY-12215] Documentation for encrypted expression support.

### PR DESCRIPTION
**Pull request by darranl**
https://issues.redhat.com/browse/WFLY-12215

### CODE REVIEW

File: docs/src/main/asciidoc/_elytron/Credential_Store.adoc [hunk 1]
```diff
@@ -3,17 +3,19 @@
 
 One new component included with WildFly Elytron for the secure storage of credentials is the Credential Store.  Previous versions of the application server made use of the Vault which was used for the secure storage of clear text strings; the credential store moves forward a step to focus on the secure storage of credentials.  As with the Vault the stored credentials could be clear text passwords however other formats are also supported.
 
-WildFly Elytron contains a default credential store implementation `KeyStoreCredentialStore` which is a store implementation backed by a `KeyStore`.
+WildFly Elytron contains two default credential store implementations.  The first implementation is the `KeyStoreCredentialStore` which is an
+implementation backed by a `KeyStore`.  The `KeyStoreCredentialStore` implementation supports the storage of various credential types,
+`PasswordCredential`, `KeyPairCredential`, and `SecretKeyCredential`.  The second implementation is the `PropertiesCredentialStore` this credential
```
Comment (bstansberry) [inline docs/src/main/asciidoc/_elytron/Credential_Store.adoc:8]: s/ this credential/. This credential/g

File: docs/src/main/asciidoc/_elytron/Credential_Store.adoc [hunk 2]
```diff
@@ -3,17 +3,19 @@
...
+store is dedicated to the storage of `SecretKeyCredential` instances using a properties file.  The `PropertiesCredentialStore` does not offer any protection
+of the credentials it stores so it's primary purpose is to provide an initial key to a server environment.
```
Comment (bstansberry) [inline docs/src/main/asciidoc/_elytron/Credential_Store.adoc:10]: s/it's/its

File: docs/src/main/asciidoc/_elytron/Credential_Store.adoc [hunk 3]
```diff
@@ -3,17 +3,19 @@
...
 
-This documentation will be focused on using the `KeyStoreCredentialStore`; however the section <<Custom_CredentialStore, Custom CredentialStore>> will describe how to convert a vault to a credential store. 
+This documentation is primarily focussed on the `KeyStoreCredentialStore` and `PropertiesCredentialStore`; however the section <<Custom_CredentialStore, Custom CredentialStore>> describes the SPIs for implementatin a custom credential store and the section <<Migrating_Existing_Vaults, Migrating Existing Vaults>> describes how to convert a vault to a credential store.
```
Comment (bstansberry) [inline docs/src/main/asciidoc/_elytron/Credential_Store.adoc:12]: https://proofreadmyessay.co.uk/writing-tips/spelling-tips-focused-or-focussed/

Up to you. :)

Comment (bstansberry) [reply docs/src/main/asciidoc/_elytron/Credential_Store.adoc:12]: s/implementatin/implementing

File: docs/src/main/asciidoc/_elytron/Credential_Store.adoc [hunk 4]
```diff
@@ -36,6 +46,8 @@ Credential Store has been successfully created
 
 This command prompts twice for the password that should be used to secure the store.  Alternatively the password can be passed in using the `--password` argument however that may mean the password is cached in the local command history or visible to other users viewing the list of running processes.
 
+NOTE: The command line tool only creates the instance of the credential store on the underlying file system, the management operation is still required to define the credential store in the `elytron` subsystem.
```
Comment (bstansberry) [inline docs/src/main/asciidoc/_elytron/Credential_Store.adoc:49]: s/system, the management/system; the management/g

File: docs/src/main/asciidoc/_elytron/Credential_Store.adoc [hunk 5]
```diff
@@ -44,31 +56,95 @@ The following operation can be used to add a `credential-store` resource to the
...
 
-=== Adding a Credential
+=== PropertiesCredentialStore / secret-key-credential-store
+
+This is a simple credential store implementation which can only be used to store `SecretKey` instances.  This credential store also does not offer any protection of the contents.  Within an application server environment it is always possible to get into a cycle of how is an initial secret provided to unlock further resources, this is primarily the purpose of this credential store.  The strategy in the past had been to use masking of a password using password based encryption, the down side of this approach was the password used for the masking was stored in the source code and being an open source project this means the password is well know.  By using the `PropertiesCredentialStore` installations are back in control of the initial secret, although this credential store does not offer it's own protection filesystem level access control can still be used to restrict access ideally to just the application server process.
```
Comment (bstansberry) [inline docs/src/main/asciidoc/_elytron/Credential_Store.adoc:77]: s/well know/well known/g

Comment (bstansberry) [reply docs/src/main/asciidoc/_elytron/Credential_Store.adoc:77]: s/initial secret, although/initial secret. Although/g

File: docs/src/main/asciidoc/_elytron/Credential_Store.adoc [hunk 6]
```diff
@@ -44,31 +56,95 @@ The following operation can be used to add a `credential-store` resource to the
...
+In this example the resource was added to reference an existing credential store, however with a small change this command can also automatically create the store if it does not exist, and
+populate it with a generated SecretKey under the alias `key` if it does not already exist.
+
+[source,options="nowrap"]
+----
+/] /subsystem=elytron/secret-key-credential-store=mycredstore:add(path=propcredstore.cs, relative-to=jboss.server.config.dir)
+{"outcome" => "success"}
+----
+
+WARNING: If you rely on automatic generation of a `SecretKey` be sure to create a backup of the key when you can, if you were to loose the key we do not have a mechanism to decrypt anything
```
Comment (bstansberry) [inline docs/src/main/asciidoc/_elytron/Credential_Store.adoc:110]: s/can, if/can. If/g

File: docs/src/main/asciidoc/_elytron/Credential_Store.adoc [hunk 7]
```diff
@@ -222,6 +378,8 @@ The following management operation can be used to remove an alias from the crede
 }
 ----
 
+By default the `credential-store` resource assumes the type to be removed is `PasswordCredential`, if a diffierent type is to be removed it can be specified with the `entry-type=SecretKeyCredential` parameter.  The `secret-key-credential-store` only holds secret keys so the entry type never needs to be specified.
```
Comment (bstansberry) [inline docs/src/main/asciidoc/_elytron/Credential_Store.adoc:381]: s/`PasswordCredential`, if/`PasswordCredential`. If/g

Comment (bstansberry) [reply docs/src/main/asciidoc/_elytron/Credential_Store.adoc:381]: s/diffierent/different

File: docs/src/main/asciidoc/_elytron/Encrypted_Expressions.adoc [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+For information regarding how to create, configure, and populate credential stores including the manipulation of secret keys please refer to the <<CredentialStore, Credential Store>> chapter.
+
+Within the `elytron` subsystem one or more `resolvers` can be defined to handle the decryption of a previously encrypted expression.  Each resolver will reference a single secret key in a
+single credential store which it will use to decrypt the expressions.  Encrypted expressions can take one of two forms:
+
+ * `${ENC::ResolverName:RUxZAUMQXUj3qP1hbXyO5PpmsbgkepjoscIf3tKXvGiPDXYqNAc=}`
+ * `${ENC::RUxZAUMQXUj3qP1hbXyO5PpmsbgkepjoscIf3tKXvGiPDXYqNAc=}`
+
+In both cases the `ENC` prefix is used to identify the expression is an encrypted expression, this prefix is the default however it is possible to define an alternative prefix for use
```
Comment (bstansberry) [inline docs/src/main/asciidoc/_elytron/Encrypted_Expressions.adoc:14]: s/expression, this/expression. This/g

File: docs/src/main/asciidoc/_elytron/Encrypted_Expressions.adoc [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+----
+[standalone@localhost:9990 /] /subsystem=elytron/expression=encryption:add(prefix=ENCRYPTED, default-resolver=one, \
+    resolvers=[{name=one, credential-store=store-one, secret-key=key}, \
+               {name=two, credential-store=store-two,secret-key=key}])
+{"outcome" => "success"}
+----
+
+== Creating Expressions
+
+NOTE: The following section illustrates how to create an encrypted expression, if you repeat the same command for the same clear text is it normal that a different
```
Comment (bstansberry) [inline docs/src/main/asciidoc/_elytron/Encrypted_Expressions.adoc:49]: s/expression, if/expression. If/g

Comment (bstansberry) [reply docs/src/main/asciidoc/_elytron/Encrypted_Expressions.adoc:49]: s/is it normal/it is normal/g

File: docs/src/main/asciidoc/_elytron/Encrypted_Expressions.adoc [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+it will need to be included in an expression as described earlier i.e. using the appropriate prefix and if required resolver name.
+
+When using the `--encrypt` action it is also possible to pass in `--clear-text` parameter to pass in the clear text directly but this may be visible to other users and may also
+be cached in the command history of your shell. 
+
+== Domain Mode
+
+When using encrypted expressions in domain mode things are slightly different to how the legacy vault may have been used in the past.
+
+To make use of encrypted expressions in the host controler configuration the `expression=encryption` resource and relevant `*credential-store` definitions must be defined within the `elytron` subsystem definition of the
```
Comment (bstansberry) [inline docs/src/main/asciidoc/_elytron/Encrypted_Expressions.adoc:112]: s/controler/controller

File: docs/src/main/asciidoc/_elytron/Encrypted_Expressions.adoc [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+
+[source,options="nowrap"]
+----
+]$ bin/elytron-tool.sh credential-store --location standalone/configuration/store-one.cs --type PropertiesCredentialStore --encrypt key
+Clear text value:
+Confirm clear text value:
+Clear text encrypted to token 'RUxZAUMQvGzk6Vaadp2cahhZ6rlPhHOZcWyjXALlAthrENvRTvQ=' using alias 'key'.
+----
+
+The `--encrypt` action is used with the `credential-store` command, the argument to this action is the alias of the secret key to use.  In this form the tool will prompt
```
Comment (bstansberry) [inline docs/src/main/asciidoc/_elytron/Encrypted_Expressions.adoc:101]: s/command, the argument/command. The argument/g

File: docs/src/main/asciidoc/_elytron/Encrypted_Expressions.adoc [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+twice for the clear text which is being encrypted.  When using the command line tool the output is just the Base64 encoded encrypted ciphertext, to use this in the management model
```
Comment (bstansberry) [inline docs/src/main/asciidoc/_elytron/Encrypted_Expressions.adoc:102]: s/ciphertext, to use/ciphertext. To use/g

**Review by bstansberry — CHANGES_REQUESTED** [review]

**Comment by darranl** [comment]
@bstansberry FYI I have marked each of your comments as resolved as I have implemented every suggestion.

**Pull request merged**